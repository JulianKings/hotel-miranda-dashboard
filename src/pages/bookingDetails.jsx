import { useState } from 'react'
import styled from 'styled-components';

export default function BookingDetails()
{
    const [count, setCount] = useState(0)

    const Button = styled.button`
        background: transparent;
        border-radius: 3px;
        border: 2px solid purple;
        color: purple;
        margin: 0.5em 1em;
        padding: 0.25em 1em;`;

    const SecondaryButton = styled.button`
    border-radius: 3px;
    border: 2px solid purple;
    background: purple;
    color: white;
    margin: 0.5em 1em;
    padding: 0.25em 1em;`

    return (
        <>
          <h1>BOOKINGS DETAILS</h1>
          <div className="card">
            <Button onClick={() => setCount((count) => count + 1)}>
              BOOKINGS DETAILS {count}
            </Button>
            <SecondaryButton onClick={() => setCount((count) => count + 1)}>
              BOOKINGS DETAILS {count}
            </SecondaryButton>
          </div>
        </>
      )
}