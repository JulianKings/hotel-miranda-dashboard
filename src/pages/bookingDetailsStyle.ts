import styled from "styled-components";

export const BookingContainer = styled.div`
  	display: grid;
	grid-template-columns: 50% 50%;

	img {
		width: 100%;
	}

	div {
		display: flex;
		flex-direction: column;
		gap: 0.75rem;
	}
`;