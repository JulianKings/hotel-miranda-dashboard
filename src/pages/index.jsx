/* eslint-disable react-hooks/exhaustive-deps */
import { BsFillHouseFill } from 'react-icons/bs';
import { IoBed } from 'react-icons/io5';
import { MdLogout, MdOutlineLogin } from 'react-icons/md';
import styled from 'styled-components';
import { Fragment, useEffect, useState } from 'react';
import singleBed from '../assets/room1.png';
import doubleBed from '../assets/room2.png';
import doubleSuperior from '../assets/room3.png';
import suite from '../assets/room4.png';
import { NavLink, useOutletContext } from 'react-router-dom';
import GuestComments from '../components/GuestComments';
import { useDispatch, useSelector } from 'react-redux';
import { fetchBookings, selectBookings, selectFetchBookingsStatus } from '../redux/slices/bookingsSlice';
import { fetchRooms, selectFetchRoomStatus, selectRooms } from '../redux/slices/roomSlice';
import { MainComponent } from '../styledcomponents/main';
import { CircularProgress } from '@mui/material';

const KPIHolder = styled.div`
	width: 100%;
	display: grid;
	grid-template-columns: repeat(auto-fill, minmax(20.25rem, 1fr));
	gap: 1.38rem;`;

const KPIItem = styled.div`
	display: flex;
	align-items: center;
	background-color: #FFFFFF;
	border-radius: 0.75rem;
	padding: 1.88rem 1.88rem;
	box-shadow: 0rem 0.25rem 0.25rem rgb(0, 0, 0, .05);

	font-size: 1.88rem;
	line-height: 2.88rem;
	color: #393939;
	font-weight: 600;
	transition: all 0.8s ease-in-out;

	&:hover {
		box-shadow: 0rem 1rem 1.88rem rgb(0, 0, 0, 0.14);	
	}

	&:hover div {
		background-color: #E23428;
		color: #FFFFFF;
	}
`;

const KPIItemImage = styled.div`
	display: flex;
	justify-content: center;
	align-items: center;
	background-color: #FFEDEC;
	width: 4.06rem;
	height: 4.06rem;
	padding: 1.13rem 1.13rem;
	color: #E23428;
	border-radius: 0.5rem;
	transition: background-color 0.6s ease-in-out, color 0.6s ease-in-out;		
`

const KPIItemText = styled.p`
	display: flex;
	flex-direction: column;
	gap: 0.25rem;
	margin-left: 1.38rem;

	span {
		display: block;
		color: #787878;
		font-weight: 300;
		font-size: 0.88rem;
		line-height: 1.31rem;

	}
`;

const InformationHolder = styled.div`
	margin-top: 2.5rem;
	display: grid;
	grid-template-columns: 47% 47%;
	grid-template-rows: 1fr;;
	gap: 2.44rem;
	justify-content: center;
	`;

const InformationBox = styled.div`
	display: flex;
	flex-direction: column;
	background-color: #FFFFFF;
	border-radius: 1.25rem;
	padding: 1.88rem 1.8rem;
	width: 100%;
	gap: 0.75rem;
	box-shadow: 0rem 0.25rem 0.25rem rgba(0, 0, 0, 0.02);
`;

const InputBox = styled.div`
	display:flex;
	flex-direction: column;
	gap: 0.25rem;
	
	input {
		background-color: #EBF1EF;
		border: 0rem solid;
		padding: 0.5rem 0.5rem;
		border-radius: 0.25rem;

		&:focus {
			outline: none;
		}
	}`;

const RoomListBox = styled.div`
	margin-top: 2.5rem;
	margin-bottom: 2.5rem;
	display: flex;
	flex-direction: column;
	gap: 1.75rem;
	background-color: #FFFFFF;
	border-radius: 1.25rem;
	padding: 1.88rem 1.8rem;
	width: 100%;
	box-shadow: 0rem 0.25rem 0.25rem rgba(0, 0, 0, 0.02);
`;

const RoomListItem = styled.div`
	width: 100%;
	display: grid;
	gap: 1.31rem;
	grid-template-columns: 9.63rem 1fr 30%;
	grid-template-rows: 1fr;
	align-items: center;
	transition: transform 0.4s ease-in-out;

	a {
		text-decoration: none;
		color: #393939;

		&:hover {
			color: #135846;
		}
	}

	&:hover {
		transform: scale(1.02);
	}

	&:hover p.room_type {
		color: #EBF1EF;
		background-color: #135846;
	}

	img {
		width: 100%;
		border-radius: 0.5rem;
	}
`;

const RoomListItemInformation = styled.div`
	display:flex;
	flex-direction: column;
	gap: 0.75rem;
	align-items: flex-start;
	color: #393939;
	font-weight: 600;
	font-size: 1.25rem;
	line-height: 1.88rem;

	p.customer_name {
		font-size: 0.88rem;
		line-height: 1.31rem;
		font-weight: 300;
	}

	p.room_type {
		padding: 0.75rem 0.75rem;
		border-radius: 0.75rem;
		background-color: #EBF1EF;
		color: #135846;
		font-weight: 300;
		font-size: 1.25rem;
		line-height: 1.88rem;
		transition: background-color 0.6s ease-in-out, color 0.4s ease-in-out;
	}
`

const RoomListItemCheckInOut = styled.div`
	display:flex;
	flex-direction: column;
	gap: 0.8rem;
	align-items: flex-end;

	p {
		padding: 0.75rem 0.75rem;
		border-radius: 0.75rem;
		background-color: #135846;
		color: #FFFFFF;
		font-weight: 300;
		font-size: 1.25rem;
		line-height: 1.88rem;
	}

	p.check_out {
		background-color: #E23428;
	}
`

const RoomListMore = styled.div`
	width: 100%;
	display: flex;
	justify-content: center;
	margin-top: 1.25rem;

	button {
		background-color: #ffffff;
		border: 0rem solid;
		color: #135846;
		font-weight: 600;
		font-size: 1rem;
		line-height: 1.56rem;
	}
`;

export default function Index()
{
	const bookingList = useSelector(selectBookings);
	const fetchBookingStatus = useSelector(selectFetchBookingsStatus);
	const roomList = useSelector(selectRooms);
	const fetchRoomStatus = useSelector(selectFetchRoomStatus);
	const dispatch = useDispatch();

	useEffect(() => {
		if(!fetchRoomStatus || !roomList || fetchRoomStatus === 'fulfilled')
		{
			dispatch(fetchRooms());
		}

		if(!fetchBookingStatus || !bookingList || fetchBookingStatus === 'fulfilled')
		{
			dispatch(fetchBookings());
		}
	}, []);

	const [startDate, updateStartDate] = useState(null);
	const [endDate, updateEndDate] = useState(null);
	const [viewMore, updateViewMore] = useState(0);
	const [sidebar] = useOutletContext();

	const [filteredBookings, updateFilteredBookings] = useState(bookingList);
	
	useEffect(() => {
		if(bookingList.length > 0)
		{
			updateFilteredBookings(bookingList);
		}
	}, [bookingList])
	
	useEffect(() => {
		if(endDate)
		{	
			if(startDate)
			{
				const filtered = bookingList.filter((booking) => new Date(booking.check_out) < endDate);
				updateFilteredBookings(filtered.filter((booking) => new Date(booking.check_in) > startDate));	
			} else {
				updateFilteredBookings(bookingList.filter((booking) => new Date(booking.check_out) < endDate))
			}

			updateViewMore(0);
		} else {
			updateFilteredBookings([...bookingList]);
		}		
	}, [startDate, endDate]);

	let scheduledRooms = 0;
	if(roomList)
	{
		scheduledRooms = Math.round((((roomList.filter((room) => room.status === 'booked')).length) / roomList.length) * 100);
	}

	return ((fetchRoomStatus !== 'fulfilled' || fetchBookingStatus !== 'fulfilled') ? <MainComponent><CircularProgress /></MainComponent> :
		<Fragment>
			<KPIHolder>
				<KPIItem>
					<KPIItemImage><IoBed size={28} /></KPIItemImage>
					<KPIItemText>
						<p>{filteredBookings.length}</p>
						<span>Bookings</span>
					</KPIItemText>
				</KPIItem>

				<KPIItem>
					<KPIItemImage><BsFillHouseFill size={28} /></KPIItemImage>
					<KPIItemText>
						<p>{scheduledRooms}%</p>
						<span>Scheduled Rooms</span>
					</KPIItemText>
				</KPIItem>

				<KPIItem>
					<KPIItemImage><MdOutlineLogin size={28} /></KPIItemImage>
					<KPIItemText>
						<p>{filteredBookings.filter((booking) => new Date(booking.check_in) > (new Date())).length}</p>
						<span>Check In</span>
					</KPIItemText>
				</KPIItem>
				
				<KPIItem>
					<KPIItemImage><MdLogout size={28} /></KPIItemImage>
					<KPIItemText>
						<p>{filteredBookings.filter((booking) => new Date(booking.check_out) < (new Date())).length}</p>
						<span>Check Out</span>
					</KPIItemText>
				</KPIItem>
			</KPIHolder>

			<InformationHolder>
				<InformationBox>
					<InputBox>
						<label htmlFor='start_date'>Start date: </label>
						<input type='date' id='start_date' onChange={(event) => {
							updateStartDate(new Date(event.target.value));
						}} />
					</InputBox>

					<InputBox>
						<label htmlFor='end_date'>End date: </label>
						<input type='date' id='end_date' onChange={(event) => {
							updateEndDate(new Date(event.target.value));
						}} />
					</InputBox>
				</InformationBox>

				<InformationBox>
					Graph
				</InformationBox>
			</InformationHolder>

			<RoomListBox>
				{
					(startDate && endDate) ? (
						filteredBookings.sort((a, b) => (new Date(a.check_in)) - (new Date(b.check_in))).slice(0, (3 + viewMore)).map((booking) => {
							return <Fragment key={booking.id}>
								<RoomListItem>
									<img src={(booking.room_type === 'Suite') ? suite : 
										(booking.room_type === 'Double Superior') ? doubleSuperior :
											(booking.room_type === 'Double Bed') ? doubleBed : singleBed
									} alt='Room Image' />
								
									<RoomListItemInformation>
										<p><NavLink to={'/booking/' + booking.id}>Room #{booking.room_number}</NavLink></p>
										<p className='customer_name'>{booking.customer_name}</p>
										<p className='room_type'>{booking.room_type}</p>
									</RoomListItemInformation>
									<RoomListItemCheckInOut>
										<p>{new Date(booking.check_in).toDateString()}</p>
										<p className='check_out'>{new Date(booking.check_out).toDateString()}</p>
									</RoomListItemCheckInOut>
								</RoomListItem>
							</Fragment>
						})
					) : 'Please select a valid date on the calendar to showcase bookings.'
				}

				{
					(endDate && startDate && (viewMore + 3) <= filteredBookings.length) ? 
					(<RoomListMore>
						<button type='button' onClick={() => {
							const increase = viewMore + 3;
							updateViewMore(increase);
						}}>View More</button>
					</RoomListMore>) : ''
				}
			</RoomListBox>

			<GuestComments sidebarStatus={sidebar}></GuestComments>
		</Fragment>
	)
}