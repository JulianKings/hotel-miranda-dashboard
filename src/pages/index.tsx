/* eslint-disable react-hooks/exhaustive-deps */
import { BsFillHouseFill } from 'react-icons/bs';
import { IoBed } from 'react-icons/io5';
import { MdLogout, MdOutlineLogin } from 'react-icons/md';
import { Fragment, useEffect, useState } from 'react';
import singleBed from '../assets/room1.png';
import doubleBed from '../assets/room2.png';
import doubleSuperior from '../assets/room3.png';
import suite from '../assets/room4.png';
import { NavLink, useOutletContext } from 'react-router-dom';
import GuestComments from '../components/GuestComments';
import { fetchBookings, selectBookings, selectFetchBookingsStatus } from '../redux/slices/bookings';
import { fetchRooms, selectFetchRoomStatus, selectRooms } from '../redux/slices/room';
import { MainComponent } from '../styledcomponents/main';
import { CircularProgress } from '@mui/material';
import { ContextType } from '../interfaces/layoutManagement';
import { useApiDispatch, useApiSelector } from '../redux/store';
import { ApiBookingInterface, ApiRoomInterface } from '../interfaces/apiManagement';
import { KPIHolder, KPIItem, KPIItemImage, KPIItemText, InformationHolder, InformationBox, InputBox, RoomListBox, RoomListItem, RoomListItemInformation, RoomListItemCheckInOut, RoomListMore } from './indexStyle';
import { DatePicker } from '@mantine/dates';

export default function Index()
{
	const bookingList: ApiBookingInterface[] = useApiSelector(selectBookings);
	const fetchBookingStatus: (string | null) = useApiSelector(selectFetchBookingsStatus);
	const roomList: ApiRoomInterface[] = useApiSelector(selectRooms);
	const fetchRoomStatus: (string | null) = useApiSelector(selectFetchRoomStatus);
	const dispatch = useApiDispatch();

	useEffect(() => {
		if(!fetchRoomStatus || !roomList || fetchRoomStatus === 'fulfilled')
		{
			dispatch(fetchRooms());
		}

		if(!fetchBookingStatus || !bookingList || fetchBookingStatus === 'fulfilled')
		{
			dispatch(fetchBookings());
		}
	}, []);

	const [startDate, updateStartDate] = useState<Date | null>(null);
	const [endDate, updateEndDate] = useState<Date | null>(null);
	const [viewMore, updateViewMore] = useState<number>(0);
	const {sidebar} = useOutletContext<ContextType>();

	const [filteredBookings, updateFilteredBookings] = useState<ApiBookingInterface[]>(bookingList);
	
	useEffect(() => {
		if(bookingList.length > 0)
		{
			updateFilteredBookings(bookingList);
		}
	}, [bookingList])
	
	useEffect(() => {
		if(endDate)
		{	
			if(startDate)
			{
				const filtered = bookingList.filter((booking) => new Date(booking.check_out) < endDate);
				updateFilteredBookings(filtered.filter((booking) => new Date(booking.check_in) > startDate));	
			} else {
				updateFilteredBookings(bookingList.filter((booking) => new Date(booking.check_out) < endDate))
			}

			updateViewMore(0);
		} else {
			updateFilteredBookings([...bookingList]);
		}		
	}, [startDate, endDate]);

	let scheduledRooms = 0;
	if(roomList)
	{
		scheduledRooms = Math.round((((roomList.filter((room) => room.status === 'booked')).length) / roomList.length) * 100);
	}

	return ((fetchRoomStatus !== 'fulfilled' || fetchBookingStatus !== 'fulfilled') ? <MainComponent><CircularProgress /></MainComponent> :
		<Fragment>
			<KPIHolder>
				<KPIItem>
					<KPIItemImage className='imageDiv'><IoBed size={28} /></KPIItemImage>
					<KPIItemText>
						<p>{filteredBookings.length}</p>
						<span>Bookings</span>
					</KPIItemText>
				</KPIItem>

				<KPIItem>
					<KPIItemImage className='imageDiv'><BsFillHouseFill size={28} /></KPIItemImage>
					<KPIItemText>
						<p>{scheduledRooms}%</p>
						<span>Scheduled Rooms</span>
					</KPIItemText>
				</KPIItem>

				<KPIItem>
					<KPIItemImage className='imageDiv'><MdOutlineLogin size={28} /></KPIItemImage>
					<KPIItemText>
						<p>{filteredBookings.filter((booking) => new Date(booking.check_in) > (new Date())).length}</p>
						<span>Check In</span>
					</KPIItemText>
				</KPIItem>
				
				<KPIItem>
					<KPIItemImage className='imageDiv'><MdLogout size={28} /></KPIItemImage>
					<KPIItemText>
						<p>{filteredBookings.filter((booking) => new Date(booking.check_out) < (new Date())).length}</p>
						<span>Check Out</span>
					</KPIItemText>
				</KPIItem>
			</KPIHolder>

			<InformationHolder>
				<InformationBox>
					<DatePicker type='range' onChange={(date) => { 
						if(date && date.length === 2 && date[0] && date[1])
						{
							updateStartDate(date[0]);
							updateEndDate(date[1]);
						}
					 }} />
				</InformationBox>

				<InformationBox>
					Graph
				</InformationBox>
			</InformationHolder>

			<RoomListBox>
				{
					(startDate && endDate) ? ((filteredBookings.length > 0) ?
						filteredBookings.slice(0, (3 + viewMore)).map((booking: ApiBookingInterface) => {
							return <Fragment key={booking._id}>
								<RoomListItem>
									<img src={(booking.room.type === 'Suite') ? suite : 
										(booking.room.type === 'Double Superior') ? doubleSuperior :
											(booking.room.type === 'Double Bed') ? doubleBed : singleBed
									} alt='Room Image' />
								
									<RoomListItemInformation>
										<p><NavLink to={'/booking/' + booking._id}>Room #{booking.room.number}</NavLink></p>
										<p className='customer_name'>{booking.client.name}</p>
										<p className='room_type'>{booking.room.type}</p>
									</RoomListItemInformation>
									<RoomListItemCheckInOut>
										<p>{new Date(booking.check_in).toDateString()}</p>
										<p className='check_out'>{new Date(booking.check_out).toDateString()}</p>
									</RoomListItemCheckInOut>
								</RoomListItem>
							</Fragment>
						}) : 'There are no bookings scheduled for the selected date range.')
					: 'Please select a valid date on the calendar to showcase bookings.'
				}

				{
					(endDate && startDate && (viewMore + 3) <= filteredBookings.length) ? 
					(<RoomListMore>
						<button type='button' onClick={() => {
							const increase: number = viewMore + 3;
							updateViewMore(increase);
						}}>View More</button>
					</RoomListMore>) : ''
				}
			</RoomListBox>

			<GuestComments $sidebarStatus={sidebar}></GuestComments>
		</Fragment>
	)
}