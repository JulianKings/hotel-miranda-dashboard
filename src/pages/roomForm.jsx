/* eslint-disable react-hooks/exhaustive-deps */
import PropTypes from 'prop-types';
import styled from 'styled-components';
import useMultiRefs from '../util/multiRef';
import { Fragment, useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRoomById, postRoom, putRoom, selectCurrentRoom, selectFetchRoomStatus } from '../redux/slices/roomSlice';
import { MainComponent } from '../styledcomponents/main';
import { CircularProgress } from '@mui/material';

export const FormButton = styled.button`
    border-radius: 0.19rem;
    border: ${props => props.buttonColor ? '0.13rem solid ' + props.buttonColor : '0.13rem solid #135846'};
    background: ${props => props.buttonColor ? props.buttonColor : '#135846'};
    color: white;
    margin: 0.75rem 0;
    padding: 0.25rem 1rem;
    width: 40%;
    max-width: 30ch;`

const FormInput = styled.input.attrs({
        type: "text",
    })`
    border: 0;
    background-color: white;
    padding: 0.45rem 0.35rem;
    border-radius: 0.25rem;
    width: 40%;
    max-width: 30ch;
    border: ${props => props.showError ? '0.16rem solid #df0000' : '0rem solid'};

    &:focus {
        outline: none;
    }
`;

const NumInput = styled.input.attrs({
    type: "number",
})`
    border: 0;
    background-color: white;
    padding: 0.45rem 0.35rem;
    border-radius: 0.25rem;
    width: 40%;
    max-width: 30ch;
    border: ${props => props.showError ? '0.16rem solid #df0000' : '0rem solid'};

    &:focus {
        outline: none;
    }
`;

const FormSelect = styled.select`
    border: 0;
    background-color: white;
    padding: 0.45rem 0.35rem;
    border-radius: 0.25rem;
    width: 40%;
    max-width: 30ch;
    border: ${props => props.showError ? '0.16rem solid #df0000' : '0rem solid'};

    &:focus {
        outline: none;
    }
`;

const FormError = styled.div`
    display: block;
    white-space: pre-line;
    background-color: #df0000;
    color: white;
    margin: 0 0.3rem;
    padding: 0.4rem 0.4rem;
    font-size: 0.8rem;
    line-height: 1.1rem;
    border-radius: 0.25rem;
    margin-bottom: 0.25rem;
    width: 40%;
    max-width: 30ch;
`

const FormBox = styled.form`
    display: flex;
    flex-direction: column;
    gap: 0.65rem;
    justify-content: center;
    align-items: center;
`;

export default function RoomForm({editMode = false})
{
    const [inputList, addInputList] = useMultiRefs();
    const [inputError, setInputError] = useState(null);
    const [inputErrorId, setInputErrorId] = useState(null);
	const navigate = useNavigate();

    const { id } = useParams();

    let roomObject = useSelector(selectCurrentRoom);
    if(!editMode)
    {
        roomObject = null;
    }
	const fetchStatus = useSelector(selectFetchRoomStatus);
	const dispatch = useDispatch();

	useEffect(() => {
		if(editMode && !roomObject || editMode && roomObject && roomObject.id !== id)
		{
			dispatch(fetchRoomById(id));
		}
	}, [id]);

    return (editMode && fetchStatus !== 'fulfilled') ? 
    <MainComponent><CircularProgress /></MainComponent>
    :
    <Fragment>
        <FormBox method='post' onSubmit={executeForm}>
            {(inputError) ? <Fragment>
                <FormError>
                    {inputError}
                </FormError>
            </Fragment> : ''}
            
            <label htmlFor='roomid'>Room Number</label>
            <NumInput id='roomid' defaultValue={(roomObject) ? roomObject.number : ''} 
						ref={addInputList}
						showError={(inputErrorId === 'roomid')} 
						onBlur={(event) => validateField(event.target)}  />

            <label htmlFor='roomprice'>Room Price</label>
            <NumInput id='roomprice' defaultValue={(roomObject) ? roomObject.price : ''}
                        ref={addInputList}
						showError={(inputErrorId === 'roomprice')} 
						onBlur={(event) => validateField(event.target)}  />

            <label htmlFor='roomdiscount'>Discount</label>
            <NumInput id='roomdiscount' defaultValue={(roomObject) ? roomObject.offer : ''}
						ref={addInputList}
						showError={(inputErrorId === 'roomdiscount')} 
						onBlur={(event) => validateField(event.target)}  />

            <label htmlFor='roomtype'>Room Type</label>
            <FormSelect ref={addInputList} id='roomtype'>
                {(roomObject && roomObject.type === 'Single Bed') ? 
                    <Fragment><option value='Single Bed' selected>Single Bed</option></Fragment> : 
                    <Fragment><option value='Single Bed'>Single Bed</option></Fragment>
                }
                {(roomObject && roomObject.type === 'Double Bed') ? 
                    <Fragment><option value='Double Bed' selected>Double Bed</option></Fragment> : 
                    <Fragment><option value='Double Bed'>Double Bed</option></Fragment>
                }
                {(roomObject && roomObject.type === 'Double Superior') ? 
                    <Fragment><option value='Double Superior' selected>Double Superior</option></Fragment> : 
                    <Fragment><option value='Double Superior'>Double Superior</option></Fragment>
                }
                {(roomObject && roomObject.type === 'Suite') ? 
                    <Fragment><option value='Suite' selected>Suite</option></Fragment> : 
                    <Fragment><option value='Suite'>Suite</option></Fragment>
                }
            </FormSelect>

            <label htmlFor='roomstatus'>Room Status</label>
            <FormSelect ref={addInputList} id='roomstatus'>                
                {(roomObject && roomObject.status === 'available') ? 
                    <Fragment><option value='available' selected>Available</option></Fragment> : 
                    <Fragment><option value='available'>Available</option></Fragment>
                }
                {(roomObject && roomObject.status === 'maintenance') ? 
                    <Fragment><option value='maintenance' selected>Maintenance</option></Fragment> : 
                    <Fragment><option value='maintenance'>Maintenance</option></Fragment>
                }
                {(roomObject && roomObject.status === 'booked') ? 
                    <Fragment><option value='booked' selected>Booked</option></Fragment> : 
                    <Fragment><option value='booked'>Booked</option></Fragment>
                }
            </FormSelect>

            <label htmlFor='roomfloor'>Room Floor</label>
            <FormInput id='roomfloor' defaultValue={(roomObject) ? roomObject.floor : ''}
						ref={addInputList}
						showError={(inputErrorId === 'roomfloor')} 
						onBlur={(event) => validateField(event.target)}  />

            <label htmlFor='roomcancellation'>Room Cancellation Policy</label>
            <FormInput id='roomcancellation' 
						ref={addInputList}
						showError={(inputErrorId === 'roomcancellation')} 
						onBlur={(event) => validateField(event.target)}  />

            <label htmlFor='roomamenities'>Room Amenities</label>
            <FormInput id='roomamenities' defaultValue={(roomObject) ? roomObject.amenities : ''}  
						ref={addInputList}
						showError={(inputErrorId === 'roomamenities')} 
						onBlur={(event) => validateField(event.target)}  />

            <label htmlFor='roompicture'>Room pictures (separated by a comma)</label>
            <FormInput id='roompicture' 
						ref={addInputList} defaultValue={(roomObject) ? roomObject.images.map( (elem) => (""+elem) ).join(',') : ''}
						showError={(inputErrorId === 'roompicture')} 
						onBlur={(event) => validateField(event.target)}  />


            <label htmlFor='roomdetails'>Room Description</label>
            <textarea ref={addInputList} id='roomdetails' cols={46} rows={6}>{(roomObject) ? roomObject.description : ''}</textarea>
            <FormButton>{(editMode) ? 'Update Room' : 'Add new Room'}</FormButton>
            {(editMode) ? <Fragment>
                <FormButton buttonColor='#df0000' type='button'
                    onClick={() => {
                        navigate('/room/' + roomObject.id + '/delete');
                    }}>Delete room</FormButton>
            </Fragment> : ''}
        </FormBox>
    </Fragment>;

    function validateField(target)
    {
        if(target)
        {
            setInputErrorId(null);
            setInputError(null);
        }
    }

    function executeForm(event)
    {
        event.preventDefault();

        const inputs = inputList();
        const inputObject = {};

        let error = false;

        inputs.forEach((input) => {
            const value = input.value;
            if(value.length < 1 && input.id !== 'roomcancellation')
            {                    
                setInputError('Please fill every field before trying to update.');
                setInputErrorId(input.id);
                error = true;
                return;
            } else if(input.id === 'roompicture' && !value.includes(','))
            {
                setInputError('Please input a valid image list.');
                setInputErrorId(input.id);
                error = true;
                return;
            } else {
                inputObject[input.id] = input.value;
            }
        })

        if(!error)
        {
            const pictures = [];
            const inputPictures = inputObject.roompicture.split(',');
            for(const pic of inputPictures)
            {
                if(pic.length > 0)
                {
                    pictures.push(pic);
                }
            }

            if(!editMode)
            {
                const roomObject = {
                    id: getRandomInt(10) + "ebb1d15-d047-" + getRandomInt(10500) + "-85c9-63c3ed856afb-" + getRandomInt(25000),
                    type: inputObject.roomtype,
                    floor: inputObject.roomfloor,
                    number: inputObject.roomid,        
                    amenities: inputObject.roomamenities,
                    images: pictures,
                    price: inputObject.roomprice,
                    offer: inputObject.roomdiscount,
                    status: inputObject.roomstatus,
                    description: inputObject.roomdetails
                }

                dispatch(postRoom(roomObject));
                navigate('/rooms');
            } else {
                const roomObject = {
                    id: id,
                    type: inputObject.roomtype,
                    floor: inputObject.roomfloor,
                    number: inputObject.roomid,        
                    amenities: inputObject.roomamenities,
                    images: pictures,
                    price: inputObject.roomprice,
                    offer: inputObject.roomdiscount,
                    status: inputObject.roomstatus,
                    description: inputObject.roomdetails
                }

                dispatch(putRoom(roomObject));
                navigate('/rooms');
            }
        }
    }

    function getRandomInt(max) {
        return Math.floor(Math.random() * max);
    }
}

RoomForm.propTypes = {
    editMode: PropTypes.bool
}