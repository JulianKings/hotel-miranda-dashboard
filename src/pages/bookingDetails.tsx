/* eslint-disable react-hooks/exhaustive-deps */
import roomImage from '../assets/room1.png';
import { useParams } from 'react-router-dom';
import { Fragment, useEffect } from 'react';
import { fetchBookingById, selectCurrentBooking, selectFetchBookingsStatus } from '../redux/slices/bookings';
import { MainComponent } from '../styledcomponents/main';
import { CircularProgress } from '@mui/material';
import { NullableApiBookingInterface } from '../interfaces/apiManagement';
import { useApiDispatch, useApiSelector } from '../redux/store';
import { BookingContainer } from './bookingDetailsStyle';

export default function BookingDetails()
{
	const { id } = useParams<string>();
    
    let bookingObject: NullableApiBookingInterface = useApiSelector(selectCurrentBooking);
    const fetchStatus: (string | null) = useApiSelector(selectFetchBookingsStatus);
	const dispatch = useApiDispatch();

	useEffect(() => {
		if(!bookingObject || bookingObject && bookingObject.id !== id)
		{
			dispatch(fetchBookingById(id));
		}
	}, [id]);
    return ((fetchStatus !== 'fulfilled' || !bookingObject || bookingObject && bookingObject.id !== id) ? 
		<MainComponent><CircularProgress /></MainComponent>
		:
        <Fragment>
			<BookingContainer>
				<div>
					<h1>Booking Details</h1>
					<p>#{bookingObject.id}</p>
					<p>Owner: {bookingObject.customer_name}</p>
					<p>Date: {new Date(bookingObject.date).toDateString()}</p>
					<p>Check In: {new Date(bookingObject.check_in).toDateString()}</p>
					<p>Check Out: {new Date(bookingObject.check_out).toDateString()}</p>
					<p>Room Number: {bookingObject.room_number}</p>
					<p>Room Type: {bookingObject.room_type}</p>
					<p>Status: {bookingObject.status}</p>
					<p>Notes: {bookingObject.notes}</p>
				</div>
				<div>
					<img src={roomImage} alt='Room Image' />
				</div>
			</BookingContainer>
        </Fragment>
      )
}