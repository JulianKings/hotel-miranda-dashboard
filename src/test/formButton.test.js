/* eslint-disable no-undef */
import { FormButton } from '../pages/roomForm';
import { render, screen } from '@testing-library/react';

test('try different button colors', async () => {
    render(
        <FormButton>default button</FormButton>,
    );

    await screen.findByRole('button');
    expect(screen.getByRole('button')).toMatchSnapshot();
});

test('try different button colors', async () => {
    render(
        <FormButton buttonColor='#df0000'>red button</FormButton>,
    );

    await screen.findByRole('button');
    expect(screen.getByRole('button')).toMatchSnapshot();
});

test('try different button colors', async () => {
    render(
        <FormButton buttonColor='blue'>blue button</FormButton>,
    );

    await screen.findByRole('button');
    expect(screen.getByRole('button')).toMatchSnapshot();
});