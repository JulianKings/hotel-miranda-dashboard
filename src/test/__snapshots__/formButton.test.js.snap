// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`try different button colors 1`] = `
<button
  class="sc-dsLQwm KTzQb"
>
  default button
</button>
`;

exports[`try different button colors 2`] = `
<button
  buttoncolor="#df0000"
  class="sc-dsLQwm gOcTTL"
>
  red button
</button>
`;

exports[`try different button colors 3`] = `
<button
  buttoncolor="blue"
  class="sc-dsLQwm bjrrqP"
>
  blue button
</button>
`;
